CDCgov/prime-simplereport
# Allow org user to be re-added after deletion

## Related Issue or Background Info

Closes https://github.com/CDCgov/prime-simplereport/issues/923

## Changes Proposed

- When a deleted user who is in the same org is added, instead of rejecting the request re-provision the user with the information from the request and send the user a new activation email where they can reconfigure their password, mfa, etc.

## Additional Information

- After discussion, it was clear that to minimize support burden the action to take when a user is re-added to an org after deletion is to wipe their credentials and put them back through the activation workflow.  See discussion on the ticket.

## Doing it Live (Testing with Okta)

In addition to the tests included, you can try test this workflow.

### On test.simplereport.gov

As of this writing, this branch is on `test` and you can try it out there.  Check https://test.simplereport.gov/api/actuator/info first to make sure this branch is still deployed.  To try it, you'll need

1. An Okta account that is an Org Admin
1. Another Okta account that is a member of the same Org

What to do:

1. Log in to `test` as the Org Admin
2. Navigate to the settings page
3. Select your other Org user, then "Remove user"
4. Choose "New user" and add a user with the same email address as the one you just removed

What to verify afterward:

1. That the user was re-added successful and their name, permissions, and facility access in the app are set to what you chose when re-adding them
2. Then, in Okta, look up the user at https://hhs-prime-admin.okta.com/admin/users and check to make sure their name is set to what you chose, that the account is has the status "Pending user action", and that the groups reflect what you chose when adding them
3. Ensure that a user activation email was received at the email address of the user re-added.  Follow the activation link and ensure you are able to set up the account password, mfa, etc.
4. Ensure user added is able to log in to SimpleReport (test) and has the chosen permissions

### Locally (because why not?)

This can be tested using real Okta accounts with some effort.  The description below uses `test` and it is assumed that groups, etc for the test environment will be used.  Prerequisites:

1. An Okta account that is an Org Admin
1. Auth token for the Org Admin User 
1. Another Okta account that is a member of the same Org
2. Okta credentials in your `application-local.yaml`
1. An Org to your local database with the same external id as the Org you are using in the test environment
2. Api Users in your local database with the same email addresses as the two Okta accounts

Then, to test:

1. In your local database, set `is_deleted` to true for the Api User you are going to attempt to re-add
2. In Okta, find the user from the previous step and Suspend them (in their user profile, use the "More Actions" drop down)
1. Run the backend app with the `okta-local` spring profile
2. Make a request to the graphql mutation `addUserToCurrentOrg` using the email of your Suspended Okta user
3. If the request is successful, you will receive an activation email (just like the initial one when creating an okta account)

The various error conditions can be tried by making the same request while the local Api User and Okta user are in various states.  Some ways to cause errors are to set the Api User `is_deleted` to false or leave the Okta user in a state other than SUSPENDED or a member of a different Organization.

## Checklist for Author and Reviewer

### UI
- ~~[ ] Any changes to the UI/UX are approved by design~~
- [ ] Any new or updated content (e.g. error messages) are approved by design 

### Testing

### Changes are Backwards Compatible
- ~~[ ] GraphQL schema changes are backward compatible with older version of the front-end~~

### Security

## Cloud
- [x] If there are changes that cannot be tested locally, this has been deployed to our Azure `test` environment for verification
